PWGJE/EMCALJetTasks/UserTasks/AliAnalysisTaskEmcalVsPhos.cxx

```diff
@@ -1041,6 +1169,145 @@ void AliAnalysisTaskEmcalVsPhos::FillClusterHistograms()
 }
 
 /*
+ * This function fills the histograms for the jet performance study.
+ */
+void AliAnalysisTaskEmcalVsPhos::FillJetPerformanceHistograms()
+{
+  TString histname;
+  
+  // If MC, get the MC event
+  const AliMCEvent* mcevent;
+  if (fGeneratorLevel) {
+    mcevent = MCEvent();
+  }
+  else {
+    return;
+  }
+  
+  // Loop through clusters, and plot M02 for each particle type
+  AliClusterContainer* clusters = GetClusterContainer(0);
+  AliClusterIterableMomentumContainer itcont = clusters->accepted_momentum();
+  const AliVCluster* clus;
+  for (AliClusterIterableMomentumContainer::iterator it = itcont.begin(); it != itcont.end(); it++) {
+      
+    clus = it->second;
+    
+    // Include only EMCal clusters (reject DCal and PHOS clusters)
+    if (!clus->IsEMCAL()) {
+      continue;
+    }
+    if (it->first.Phi_0_2pi() > fgkEMCalDCalPhiDivide) {
+      continue;
+    }
+    
+    // If MC, determine the particle type
+    // Each detector-level cluster contains an array of labels of each truth-level particle contributing to the cluster
+    ParticleType particleType1 = kUndefined;
+
+    Int_t label = TMath::Abs(clus->GetLabel()); // returns mc-label of particle that deposited the most energy in the cluster
+    if (label > 0) { // if the particle has a truth-level match, the label is > 0
+      
+      // Method 1: Use AliMCAnalysisUtils to identify all particles
+      particleType1 = GetParticleType1(clus, mcevent, clusters->GetArray());
+
+    }
+
+    // (M02, Eclus, part type)
+    histname = "JetPerformance/hM02VsParticleType";
+    fHistManager.FillTH3(histname, clus->GetM02(), clus->GetNonLinCorrEnergy(), particleType1);
+
+  }
+  
+  // Loop through jets, to fill various histograms
+  AliJetContainer* jets = GetJetContainer(0); // there is only a single, det-level jet finder here
+  for (const auto jet : jets->accepted()) {
+    
+    Double_t jetPt = GetJetPt(jet, 0);
+    
+    // Keep track of hadronic calo energy in each jet
+    Double_t hadCaloEnergyUnmatched = 0;
+    Double_t hadCaloEnergyMatchedNonlincorr = 0;
+    Double_t hadCaloEnergyMatchedHadCorr = 0;
+    
+    // Loop through clusters in each jet, to plot several histograms
+    Int_t nClusters = jet->GetNumberOfClusters();
+    for (Int_t iClus = 0; iClus < nClusters; iClus++) {
```

Comment:
i think you can use `for(auto cluster : clusCont->accepted())`

Reply:
Here I loop through only those clusters in the jet, not all clusters in the event, so I think it is OK as written.

Reply:
yes, you are right! i missed this (and a nice addition when i have time would be to have the jet return a container of its cluster in order to be easier iterable (and maybe even searchable))